### ISSUE TITLE
Stack overflow with large input

### ISSUE THREAD

**Issue by ruckustboom** [sargunv/cakeparse#12]
I may be doing something wrong, but I get a stack overflow when I do the following:

``` kotlin
val number = token("number", "[0-9]+")
val space = token("space", "\\s+")
val goal = oneOrMore(number before optional(space))

val input = "557196 56123\n".repeat(5000)
println(input)
val lexed = setOf(number, space).lexer().lex(input).cached()
println("lexed")
val result = lexed.parseToEnd(goal).value
println("Result: $result")
```

**Comment by sargunv** [comment]
Will investigate.

**Comment by ruckustboom** [comment]
Upon further testing, it breaks with as low as 400 repititions

**Comment by ruckustboom** [comment]
Sorry, I didn't notice you already had #7 open. This probably should have gone there.

**Comment by sargunv** [comment]
No problem, I'll close that issue once I add a (failing) test case and use this issue to continue this.

Thanks for the bug report.

**Comment by ruckustboom** [comment]
Sounds good

**Comment by sargunv** [comment]
The problem is with my recursive implementation of `atLeast(n: Int, target: Parser<A>)`. It needs to be changed to either use tail recursion or an iterative solution.

**Comment by sargunv** [comment]
Released version 1.1.0 with this fix. Let me know if you have further issues.
